File: python/linkedlist/ReverseLinkedList.py
'''
https://leetcode.com/problems/reverse-linked-list/
Given the head of a singly linked list, reverse the list, and return the reversed list.
Input: head = [1,2,3,4,5]
Output: [5,4,3,2,1]
'''
from typing import Optional
from ListNode import ListNode
from LinkedListUtils import build_linked_list, print_linked_list

class ReverseLinkedList:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev, curr = None , head
        while curr:
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next
        return prev

if __name__ == "__main__":
    # Input: head = [1,2,3,4,5]
    input_values = [1, 2, 3, 4, 5]
    head = build_linked_list(input_values)
    
    print(f"Original List: {print_linked_list(head)}")
    
    solution = ReverseLinkedList()
    reversed_head = solution.reverseList(head)
    
    # Output: [5,4,3,2,1]
    print(f"Reversed List: {print_linked_list(reversed_head)}")

File: src/main/java/com/naresh/a_dsalgo/linkedlist/implementation/problems/ReverseLinkedList.java
package com.naresh.a_dsalgo.linkedlist.implementation.problems;

/**
 * Problem: Reverse Linked List
 * Description: Reverse a singly linked list.
 */
public class ReverseLinkedList {

    static class ListNode {
        int val;
        ListNode next;

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    /**
     * Algorithm: Iterate through the list, changing the next pointer of each node
     * to point to its previous node. Use prev, curr, and nextTemp pointers.
     */
    public ListNode reverseList(ListNode head) {
        // Pattern: Iterative | Time: O(n), Space: O(1)
        ListNode prev = null;
        ListNode curr = head;
        while (curr != null) {
            ListNode nextTemp = curr.next; // Store next node
            curr.next = prev;              // Reverse pointer
            prev = curr;                   // Advance prev
            curr = nextTemp;               // Advance curr
        }
        return prev;
    }
    // FAANG Tip: Iterative approach is O(1) space, making it superior to recursion (O(n) stack) for large lists. Mention the "three-pointer" shuffle as the core mechanic.

    public static void main(String[] args) {
        var sol = new ReverseLinkedList();
        var head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
        print(sol.reverseList(head)); // Expected: 5 -> 4 -> 3 -> 2 -> 1
    }

    private static void print(ListNode node) {
        while (node != null) {
            System.out.print(node.val + (node.next != null ? " -> " : ""));
            node = node.next;
        }
        System.out.println();
    }
       /*
    1.
prev = curr;: We are updating the prev pointer to point to the node that was just the "current" one. So, prev moves to the curr position.
2.
curr = nextTemp;: We are updating the curr pointer to point to the next node in the original sequence (which we saved in nextTemp). So, curr moves to the next position.
     */
}
